### DISCUSSION TITLE
Loading Inline Scripts on unique pages

### DISCUSSION THREAD

**Discussion by byjacobbyers** [gatsbyjs/gatsby#33061] [Help]
### Preliminary Checks

- [X] This issue is not a duplicate. Before opening a new issue, please search existing issues: https://github.com/gatsbyjs/gatsby/issues
- [X] This issue is not a question, feature request, RFC, or anything other than a bug report directly related to Gatsby. Please post those things in GitHub Discussions: https://github.com/gatsbyjs/gatsby/discussions

### Description

🤔 Ran into an interesting problem and feel that I am missing something simple. 

My client gave me a script to load content for a Terms of Use Page instead of using the CMS. So I implemented it like this. 

```
import * as React from "react"




// markup
const IssuePage = () => {
  return (
    <main >
      <title>Issue Page</title>
      <h1 >
        Issue Page
      </h1>
      <div id="otnotice-9438532b-0504-4faa-882e-923bafea1b14" class="otnotice"></div>

      <script 
        src="https://appds8093.blob.core.windows.net/privacy-notice-scripts/otnotice-1.0.min.js" 
        type="text/javascript" 
        charset="UTF-8" 
        id="otprivacy-notice-script"
        dangerouslySetInnerHTML={{
          __html: `
          settings="eyJjYWxsYmFja1VybCI6Imh0dHBzOi8vcHJpdmFjeXBvcnRhbC5vbmV0cnVzdC5jb20vcmVxdWVzdC92MS9wcml2YWN5Tm90aWNlcy9zdGF0cy92aWV3cyJ9"
        `,
        }}
      />
      <script 
        type="text/javascript" 
        charset="UTF-8"
        dangerouslySetInnerHTML={{
          __html: `
          OneTrust.NoticeApi.Initialized.then(function() {
            OneTrust.NoticeApi.LoadNotices(["https://appds8093.blob.core.windows.net/3546d798-21cd-434d-b031-f8701f4d2f21/privacy-notices/9438532b-0504-4faa-882e-923bafea1b14.json"], undefined, undefined, false);
          });
        `,
        }}
      />
    </main>
  )
}

export default IssuePage
```

The content does not show on the initial load but works after I refresh the page. I am not sure what is preventing it load in the first place and thought that this could be an issue. Mainly because I can see people wanting to have a simple solution for using inline scripts on unique pages and not use gatsby-ssr or index.js for site-wide scripts. 

### Reproduction Link

https://github.com/byebyers/inlinejsrepro

### Steps to Reproduce

1. clone the repo and install packages
2. gatsby build 
3. gatsby serve
3. navigate to the issue page to see no content showing
4. refresh the page to see content
...

### Expected Result

Content should show

### Actual Result

I have to refresh the page to see content

### Environment

```shell
System:
    OS: Linux 5.10 Ubuntu 20.04.2 LTS (Focal Fossa)
    CPU: (8) x64 Intel(R) Core(TM) i7-6700K CPU @ 4.00GHz
    Shell: 5.0.17 - /bin/bash
  Binaries:
    Node: 14.17.4 - ~/.nvm/versions/node/v14.17.4/bin/node
    npm: 7.21.1 - ~/.nvm/versions/node/v14.17.4/bin/npm
  npmPackages:
    gatsby: ^3.13.0 => 3.13.0
    gatsby-plugin-image: ^1.13.0 => 1.13.0
    gatsby-plugin-manifest: ^3.13.0 => 3.13.0
    gatsby-plugin-mdx: ^2.13.0 => 2.13.0
    gatsby-plugin-netlify-cms: ^5.13.0 => 5.13.0
    gatsby-plugin-postcss: ^4.13.0 => 4.13.0
    gatsby-plugin-react-helmet: ^4.13.0 => 4.13.0
    gatsby-plugin-sharp: ^3.13.0 => 3.13.0
    gatsby-plugin-sitemap: ^4.9.0 => 4.9.0
    gatsby-source-filesystem: ^3.13.0 => 3.13.0
    gatsby-transformer-sharp: ^3.13.0 => 3.13.0
  npmGlobalPackages:
    gatsby-cli: 3.10.0
```

**Comment by byjacobbyers** [comment]
I was able to get around this issue by adding the script & content to an HTML page and accessing it via Gatsby's static folder.

**Comment by YungTosti** [comment]
Is there a solid solution for this problem yet? I've been stuck on this for quite some time.

**Comment by byjacobbyers** [comment]
Hey @YungTosti, I have 2 solutions that may work.

First solution you can try: [dangerously-set-html-content](https://www.npmjs.com/package/dangerously-set-html-content) from npm

Another is to use useState & useEffect on your page. Which is based on this [Stack Overflow Answer](https://stackoverflow.com/questions/56943000/adding-marketo-form-to-gatsby-jsx)

```
import React, { useState, useEffect } from 'react';

const scriptId= 'scriptId'; //in string format

const Component = ({  props }) => {
  const [isLoaded, setIsLoaded] = useState(false);

  useEffect(() => {
    if (!document.getElementById(scriptId)) {
      loadScript();
      // eslint-disable-next-line react-hooks/exhaustive-deps
    } else {
      setIsLoaded(true);
    }
  }, []);

  const loadScript = () => {
    var s = document.createElement('script');
    s.id = scriptId;
    s.type = 'text/javascript';
    s.src = 'https://www.pathtoyourscript.com/file.js';
    s.onreadystatechange = function() {
      if (this.readyState === 'complete' || this.readyState === 'loaded') {
        setIsLoaded(true);
      }
    };
    s.onload = () => setIsLoaded(true);
    document.getElementsByTagName('head')[0].appendChild(s);
  }; 

  return (
    <div></div>
  );
};

export default Component 

```

Hope this helps
